### ISSUE TITLE
Class Imbalance

### ISSUE THREAD

**Issue by luke2997** [vqdang/hover_net#36]
Hello, I notice there is a low miscellaneous F1 score due to class imbalance - How was class imbalance dealt with in this paper - did augmentation help? Also where and how can i change each of the loss functions in this code?

**Comment by vqdang** [comment]
Hi, we counter class balance by applying dice loss and we weight it equally with the standard cross entropy. You can read more about the loss and you can modify it here

https://github.com/vqdang/hover_net/blob/4e08189ff8eeb77cde15d0876d7e853a6ddc759e/src/model/graph.py#L266-L304

**Comment by luke2997** [comment]
Thank you - however when i try to change the DICE loss, the output images in inference are not classified it seems. 

Below I added this into graph.py:

```
def focal_loss(pred,true,alpha,gamma):
                zeros = tf.zeros_like(pred,dtype=pred.dtype)
                pos_p_sub = tf.where(true > zeros, true - pred, zeros)
                neg_p_sub = tf.where(true > zeros, zeros, pred)
                loss = - alpha * (pos_p_sub ** gamma) * tf.log(tf.clip_by_value(pred, 1e-8, 1.0)) \
                              - (1 - alpha) * (neg_p_sub ** gamma) * tf.log(tf.clip_by_value(1.0 - pred, 1e-8, 1.0))

                return tf.reduce_sum(loss)
```

And then from lines 266:

``` 
elif term == 'focal':
                    term_loss=focal_loss(pred_hv, true_hv, alpha=0.25,gamma=2)
                    term_loss=tf.identity(term_loss,name='focal')

```

I also I added this into hover.py:

``` 
  'loss_term' : {'bce' : 1,'focal' : 1, 'mse' : 2, 'msge' : 1},

```

I only had one warning when training ` /loader/augs.py:211: UserWarning: Only one label was provided to `remove_small_objects`. Did you mean to use a boolean array? ` and inference had no errors/warnings. When running compute_stats, this then clearly doesn't work for either segmentation or classification as something has gone wrong in training I believe using this loss function. Have i missed anything or should I be making changes anywhere else in the repository to use this?

**Comment by simongraham** [comment]
Have you visualised the focal loss during training? This is done by `add_moving_summary()`. Also it seems that you are applying focal loss on the HV predictions and outputs. I don't think that makes sense here.

**Comment by luke2997** [comment]
Ah yes, you're right I should be adding a prediction tensor and target tensor as these arguments, which should be of size (batch_size, number_of_anchors, number_of_anchors).

I'm unfamiliar with tensorflow, looking through your code would these just be soft_np/one_np or are they defined elsewhere? i.e.:
``` 
term_loss=focal_loss(soft_np[...,0],one_np[...,0],alpha=0.25,gamma=2)\
                    +term_loss=focal_loss(soft_np[...,1], one_np[...,1], alpha=0.25,gamma=2)
term_loss=tf.identity(term_loss,name='loss-focal')
```

And then for  `add_moving_summary()` I'd use soft_class as per DICE.
